mobile/lib/modules/album/views/album_viewer_page.dart

```diff
@@ -33,9 +36,12 @@ class AlbumViewerPage extends HookConsumerWidget {
   Widget build(BuildContext context, WidgetRef ref) {
     FocusNode titleFocusNode = useFocusNode();
     final album = ref.watch(albumWatcher(albumId));
+    // Listen provider to prevent autoDispose when navigating to other routes from within the viewer page
```

Comment:
is this a workaround for the new riverpod annotation?

Reply:
Not exactly. There is a problem with the current `StateProvider` (non-generated) implementation in main. Before code-generation, all the providers were always kept alive and never auto disposed unless explicitly opt-in. This results in an error where once you open an album, the `currentAlbumProvider` will cache the value of the current album, even after you leave the album page. This results in, for example, displaying the activity icon in the main timeline since the `currentAlbumProvider` still has a non-null value. 

Now this can be handled in two ways. One is for the provider to be disposed so that it'll reset the state and go back to null, or the value should be updated to null once we pop-out of the album viewer page. It is much easier to let the provider auto-dispose and do its cleanup than to keep track of updating the state. Since after code-generation, all the providers are autoDisposed by default, this problem of the state having a non-null value forever is fixed. AutoDispose introduced one more error after the refactoring, i.e, Once we navigate from the `album_viewer_page` -> `activities_page`, the provider was getting auto disposed since no listeners were active in the `album_viewer_page`. Using a manual `ref.listen` with an empty callback will prevent riverpod from auto-disposing the provider below the current widgets sub-tree. Once the `album_viewer_page` is popped out of the route stack, the provider will be auto disposed and will be reset to null. Using ref.listen has no performance impact other than preventing the provider from getting autoDisposed and will also not result in a widget rebuild.

We have a similar handling for the newly added `currentAssetProvider` as well. So basically, any temporary provider within a particular widget sub-tree needs either handling. The `currentUserProvider` do not need this because it is marked as `keepAlive` and is needed for the entirety of the app life cycle

Reply:
We do, however, require a workaround to use the new providers generated with the annotation for testing.

If we want to mock the notifier class and not just the service / state, then we need to extend the private generated class directly in the provider file.

https://riverpod.dev/docs/essentials/testing#mocking-notifiers

I personally, did not like having my Mock class in the same file as that of the normal provider and so, added an abstract class extending the internal class in the provider file as such
https://github.com/immich-app/immich/blob/cf76a5cf924d20e9f2f0487d548dc6a0ac575f57/mobile/lib/modules/asset_viewer/providers/current_asset.provider.dart#L14-L15

and then used that for defining the actual mock class
https://github.com/immich-app/immich/blob/cf76a5cf924d20e9f2f0487d548dc6a0ac575f57/mobile/test/modules/asset_viewer/asset_viewer_mocks.dart#L5-L7

If we are okay with adding the mock class directly in the `current_asset.provider.dart` file, then we can do it so or leave them be and keep the testing related mock class separated under the test package and non-test related code in the lib package.

Reply:
Thank you for the detailed explanation! I really appreciate it.